### DISCUSSION TITLE
opendds high cpu usage

### DISCUSSION THREAD

**Discussion by zhuohuawang** [OpenDDS/OpenDDS#4257] [Q&A] [answered]
I created five different publishers in one process, each publishing at a rate of 5ms. Then in another process, five subscribers are created for each of the five publishers to receive messages, opening a total of five processes. That is, there are five publishers, each with five subscribers.
After running, the publisher process has a 30% cpu usage and each subscriber has a 5% cpu usage, so the total CPU usage is 60%.
CPU：Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz   2.80 GHz
The configuration is as follows：
[common]
DCPSGlobalTransportConfig=$file
DCPSDefaultDiscovery=DEFAULT_RTPS
[config/transports_config]
transports=test_transport
[transport/test_transport]
transport_type=rtps_udp

Is the cpu usage normal? Whether the cpu usage can be reduced by setting parameters.

**Comment by jrw972** [comment]
> Is the cpu usage normal?

Impossible to say given that "normal" is not well defined.  Every application, platform, etc. is different.  I want to come back to this question.

> Whether the cpu usage can be reduced by setting parameters.

Probably.  I suggest you look at the QoS for the writers and readers and see if there are places where simpler alternatives can be used.  For example, turning off reliability in the writer.

# What does the OpenDDS Community need for performance testing and tuning?

Performance-related questions like this are quite common, and it would be nice if we could direct the user to a set of resources that put them on the path to understanding and improving the performance.  Thus, my question for the community is, what does the community need for performance testing and tuning?  I'll list some things to get the discussion started, but I'm very much interested in the thoughts of the community.  (I don't want to put a damper on the discussion from the start, but it is important to consider feasibility, maintainability, and cost when proposing an idea.)

Should the [DevGuide](https://opendds.readthedocs.io/en/latest-release/index.html) or some other resource have performance testing and tuning content?  What is a brief outline?  How should we expand the existing [Bench](https://opendds.readthedocs.io/en/latest-release/internal/bench.html) documentation?

It is impossible to give satisfactory answers to performance-related questions given the amount of variability in applications, IDL, QoS, compilers, platforms, networks, etc.  However, a standardized set of benchmarks would eliminate variability arising from the application. 

OpenDDS has a framework for performance testing called [Bench](https://opendds.readthedocs.io/en/latest-release/internal/bench.html).  Bench is used in a few of the standard integration tests.  Should we extend those tests to report on the measured performance?  Should the suite of tests be expanded?  What is reasonable for tests executing on a single machine?

With a standardized set of performance tests, I think the next step would be to share these results so that users and OpenDDS developers can see their platform compared to others.  To be successful, there would need to be 1) enough people running the tests often enough to create a meaningful baseline and 2) a mechanism for sharing the test results.  We can address the sharing problem in a number of ways, but the ultimate choice needs to be one that is easy to use and sustainable for everyone.  So, are users willing to share performance test results?  If so, we could expand this to include bigger scenarios, as was done in the Bench scoreboard that existed before OpenDDS Foundation transitioned to Unity.

Alternatively, we could also offer performance testing as a service, i.e., you pay to see results from generic or custom performance tests.

What are your thoughts?  What resources do user's of OpenDDS need to analyze and tune the performance of their applications?

**Reply by zhuohuawang** [reply]
Do you have some test data about resource occupancy rate that you can provide for my reference? Another issue I have found is that when I call the publish function to publish data over and over again, memory increases. If a subscriber subscribes at this point, CPU usage increases as memory increases. I can provide my code and configuration if you want.

**Reply by jrw972** [reply]
I do not have any resource occupancy data.  If memory increases with publishing, then either 1) the application is creating a new instance each time you publish or 2) the history/durability QoS is causing the sample to be saved.  If you would like to share your code, logs, observations, etc. that will certainly make it easier to suggest things to try.

**Reply by zhuohuawang** [reply]
I'm sure it's the increase in memory caused by calling the Messenger:: MessageDataWriter:: write function, because there was no increase in memory when I commented it out.
I set only WriterQos and use the default values for all other Qos. WriterQos is as follows:
writer_qos.durability.kind = DDS::VOLATILE_DURABILITY_QOS;
    // writer_qos.resource_limits.max_samples = 100;
    // writer_qos.resource_limits.max_instances = 100;
    // writer_qos.resource_limits.max_samples_per_instance = 100;
    // writer_qos.history.kind = DDS::KEEP_LAST_HISTORY_QOS;
    // writer_qos.history.depth = 100;
    writer_qos.durability.kind = DDS::VOLATILE_DURABILITY_QOS;
    writer_qos.durability_service.service_cleanup_delay.nanosec = 1000* 100;
    writer_qos.durability_service.history_depth = 100;
    writer_qos.durability_service.history_kind = DDS::KEEP_LAST_HISTORY_QOS;
    writer_qos.durability_service.max_instances = 100;
    writer_qos.durability_service.max_samples = 100;
    writer_qos.durability_service.max_samples_per_instance = 100;

ini file configuration is as follows:
[common]
DCPSGlobalTransportConfig=transports_config
DCPSDefaultDiscovery=DEFAULT_RTPS
[config/transports_config]
transports=test_transport
[transport/test_transport]
transport_type=rtps_udp

In the process of my testing I encountered another problem. If I do this with writer_qos.resource_limits.max_instances = 100, I get an error:
(16380|17868) NOTICE: Messenger::MessageDataWriterImpl::write_w_timestamp: register failed: Out of resources
(16380|17868) ERROR: DataWriterImpl::register_instance_i: register instance with container failed, returned <Out of resources>.

What is this instance? I don't see which one it refers to in the documentation, so I can't understand what max_instances are used to limit.Why an error occurs when I limit max_instances.

**Reply by iguessthislldo** [reply, marked as answer]
> In the process of my testing I encountered another problem. If I do this with writer_qos.resource_limits.max_instances = 100, I get an error:
(16380|17868) NOTICE: Messenger::MessageDataWriterImpl::write_w_timestamp: register failed: Out of resources
(16380|17868) ERROR: DataWriterImpl::register_instance_i: register instance with container failed, returned .
>
>What is this instance? I don't see which one it refers to in the documentation, so I can't understand what max_instances are used to limit.Why an error occurs when I limit max_instances.

An instance in this case is a [DDS sample instance](https://opendds.readthedocs.io/en/dds-3.25/devguide/introduction.html#topic). If a topic type has `@key` fields, every sample published by a DataWriter with a different combination of values of the `@key` fields will create a new sample instance.

For example if a writer's IDL is:

```
@topic
struct Entity {
  @key string id;
  int32 status;
};
```

Then samples `{"a", 0}` and `{"a", 100}` would be part of a single instance where `id` is `"a"` and the sample `{"b", 10}` would be part of another instance where `id` is `"b"`. Instances are registered automatically when a write happens. They can also be manually registered and also unregistered.

The [`resource_limits`](https://opendds.readthedocs.io/en/dds-3.25/devguide/quality_of_service.html#resource-limits) you're setting is limiting the number of instances your writer can handle at once. You can do one of the following:

1. If you want to use instances, don't set `max_instances`. You could continue to set it if you really want to, but you'd have keep track of instances already registered and unregister one of them when you get `OUT_OF_RESOURCES` so you can register a new one.
3. If you don't want to use instances, remove `@key`s on the IDL so you only ever have 1 instance. I'm assuming you're not using instances, so this is what I recommend.

**Comment by zhuohuawang** [comment]
Thanks for your reply. According to your prompt, I removed the @key field in the topic type, and then the memory no longer grew.

Now I have another problem.
I created two writers for two different themes in the same domain. Two threads are then used to publish data using each writer.
Records are made in functions on_subscription_matched, on_liveliness_changed and on_data_available at the receiving end. Both topics trigger on_subscription_matched, but only one of them triggers on_liveliness_changed. Two copies of data are received in on_data_available, but both are received by the subject that triggers on_liveliness_changed, which means that the subject receives a copy of data that does not belong to it.

Code to create publisher：
```
std::vector<std::thread> thrds(2);
Publisher pub;
for(size_t i =0;i < thrds.size();++i){
        std::string topic_name = "test_topic_" + std::to_string(i+1);
        uint32_t domian_id = 10;
        pub.create_publisher(domian_id,topic_name);
}
    std::atomic_bool quit = false;
    for(size_t i =0;i < thrds.size();++i){
        std::thread t([&quit, i, &pub](){
            std::string topic_name = "test_topic_" + std::to_string(i+1);
            uint32_t domian_id = 10;
            Messenger::Message message;
            message.text = "sdafgdfasdasda";
            for(int index = 0;(index < 100000 && !quit);++index){
                message.index = index;
                message.timestamp = getTimestamp();
                pub.publish(domian_id,topic_name,true,message);
                std::this_thread::sleep_for(std::chrono::milliseconds(5 * 1000));
            }
        });
        thrds[i].swap(t);}
```

```
bool Publisher::create_publisher(int32_t domain_id, std::string topic_name){
    //Avoid creating the same topic in the same domain repeatedly.
    if(writers_.count(get_writers_key(domain_id, topic_name))){
        return true;
    }
    //Use the same participant for the same domain.
    if(!participant_)
    {
        participant_ = dpf_->create_participant(domain_id,PARTICIPANT_QOS_DEFAULT, nullptr,OpenDDS::DCPS::DEFAULT_STATUS_MASK);
        // Register TypeSupport (Messenger::Message)
        ts_ = new Messenger::MessageTypeSupportImpl;
        if (ts_->register_type(participant_, "") != DDS::RETCODE_OK) {
            std::cout<<"register type error."<<std::endl;
            dpf_->delete_participant(participant_);
            return false;
        }
    }
    if(!participant_){
        std::cout<<"create participant failed."<<std::endl;
        return false;
    }
    
    // Create Topic (Movie Discussion List)
    CORBA::String_var type_name = ts_->get_type_name();
    DDS::Topic_var topic =
        participant_->create_topic(topic_name.data(),
        type_name,
        TOPIC_QOS_DEFAULT,
        0,
        OpenDDS::DCPS::DEFAULT_STATUS_MASK);
    if(!topic){
        std::cout<<"create topic failed."<<std::endl;
        return false;
    }
    // Create Publisher
    DDS::Publisher_var publisher =
    participant_->create_publisher(PUBLISHER_QOS_DEFAULT,
        0,
        OpenDDS::DCPS::DEFAULT_STATUS_MASK);
    
    if(!publisher){
        std::cout<<"create publisher failed."<<std::endl;
        return false;
    }
    DDS::DataWriterQos writer_qos;
    publisher->get_default_datawriter_qos(writer_qos);
    writer_qos.durability.kind = DDS::VOLATILE_DURABILITY_QOS;
    writer_qos.resource_limits.max_samples = 100;
    // writer_qos.resource_limits.max_instances = 100;
    writer_qos.resource_limits.max_samples_per_instance = 100;
    writer_qos.history.kind = DDS::KEEP_LAST_HISTORY_QOS;
    writer_qos.history.depth = 100;
    // Create DataWriter
    DDS::DataWriter_var writer =
    publisher->create_datawriter(topic,
        writer_qos,
        0,
        OpenDDS::DCPS::DEFAULT_STATUS_MASK);

    if (!writer) {
        std::cout<<"create writer failed."<<std::endl;
        return false;
    }
    writers_[get_writers_key(domain_id, topic_name)] = writer;
    std::cout<<"create topic "<< topic_name <<"  domain id:"<< domain_id<<std::endl;
    return true;
}

```
```
bool Publisher::publish(int32_t domain_id, std::string topic_name, bool wait_matched, const Messenger::Message &message){
    std::string key = get_writers_key(domain_id, topic_name);
    if(!writers_.count(key)){
        std::cout<<"there is no dimain/topic created."<<std::endl;
        return false;
    }
    auto writer = writers_[key];

    while (wait_matched) {
        DDS::StatusCondition_var condition = writer->get_statuscondition();
        condition->set_enabled_statuses(DDS::PUBLICATION_MATCHED_STATUS);
        DDS::WaitSet_var ws = new DDS::WaitSet;
        ws->attach_condition(condition);
        DDS::PublicationMatchedStatus matches;
        if (writer->get_publication_matched_status(matches) != ::DDS::RETCODE_OK) {
            ACE_ERROR_RETURN((LM_ERROR,
            ACE_TEXT("ERROR: %N:%l: main() -")
            ACE_TEXT(" get_publication_matched_status failed!\n")),
            1);
        }

        if (matches.current_count >= 1) {
            break;
        }

        DDS::ConditionSeq conditions;
        DDS::Duration_t timeout = { 60, 0 };
        if (ws->wait(conditions, timeout) != DDS::RETCODE_OK) {
            ACE_ERROR_RETURN((LM_ERROR,
            ACE_TEXT("ERROR: %N:%l: main() -")
            ACE_TEXT(" wait failed!\n")),
            1);
        }
    }
    
    if(!message_writer_){
        message_writer_ = Messenger::MessageDataWriter::_narrow(writer);
    }
    //Messenger::Message message;
    //message.index = 1;
    //message.text = "sdagfjhdlasjfka";
    //message.timestamp = getTimestamp();
    DDS::ReturnCode_t error = message_writer_->write(message, DDS::HANDLE_NIL);
    return error == DDS::RETCODE_OK;
 }
```


Code to create subscriber：

```
Subscriber sub1(ini_path);
    std::string topic_name = "test_topic_" + std::to_string(1);
    sub1.create_subscriber(10 ,topic_name);
    //Subscriber sub2(ini_path);
    topic_name = "test_topic_" + std::to_string(2);
    sub1.create_subscriber(10 ,topic_name);`

bool Subscriber::create_subscriber(int32_t domain_id, std::string topic_name){
    if(!participant_){
        participant_ = dpf_->create_participant(domain_id,PARTICIPANT_QOS_DEFAULT, nullptr,OpenDDS::DCPS::DEFAULT_STATUS_MASK);
        if(!participant_){
            std::cout<<"create participant failed."<<std::endl;
            return false;
        }
        // Register TypeSupport (Messenger::Message)
        ts_ = new Messenger::MessageTypeSupportImpl;
        if (ts_->register_type(participant_, "") != DDS::RETCODE_OK) {
            std::cout<<"register type error."<<std::endl;
            dpf_->delete_participant(participant_);
            return false;
        }
    }
    // Create Topic (Movie Discussion List)
    CORBA::String_var type_name = ts_->get_type_name();
    DDS::Topic_var topic =
        participant_->create_topic(topic_name.data(),
        type_name,
        TOPIC_QOS_DEFAULT,
        0,
        OpenDDS::DCPS::DEFAULT_STATUS_MASK);
    if(!topic){
        std::cout<<"create topic failed."<<std::endl;
        return false;
    }
        // Create Subscriber
    DDS::Subscriber_var subscriber =
    participant_->create_subscriber(SUBSCRIBER_QOS_DEFAULT,
        0,
        OpenDDS::DCPS::DEFAULT_STATUS_MASK);

    if (!subscriber) {
        std::cout<<"create subscriber failed."<<std::endl;
        return false;
    }
    // Create DataReader
    DDS::DataReaderListener_var listener(new DataReaderListenerImpl);
    auto lisn = dynamic_cast<DataReaderListenerImpl *>(listener.in());
    lisn->startLog(topic_name);

    DDS::DataReaderQos reader_qos;
    subscriber->get_default_datareader_qos(reader_qos);
    reader_qos.history.kind = DDS::KEEP_LAST_HISTORY_QOS;
    reader_qos.history.depth = 100;
    reader_qos.resource_limits.max_samples = 100;
    DDS::DataReader_var reader = subscriber->create_datareader(topic, reader_qos,listener,OpenDDS::DCPS::DEFAULT_STATUS_MASK);
    if(!reader){
        std::cout<<"create datareader failed."<<std::endl;
        return false;
    }
    std::cout<<"create topic "<< topic_name <<"  domain id:"<< domain_id<<std::endl;
    return true;
}

void DataReaderListenerImpl::on_liveliness_changed(
  DDS::DataReader_ptr reader,
  const DDS::LivelinessChangedStatus& status)
{
  std::cout<<"liveliness changed.topic:"<<reader->get_topicdescription()->get_name()
  <<" alive_count="<<status.alive_count<<" alive_count_change="<< status.alive_count_change
  << " not_alive_count="<<status.not_alive_count<<std::endl;
}

void DataReaderListenerImpl::on_data_available(DDS::DataReader_ptr reader)
{
  Messenger::MessageDataReader_var reader_i =
    Messenger::MessageDataReader::_narrow(reader);
  
  if (!reader_i) {
    ACE_ERROR((LM_ERROR,
               ACE_TEXT("ERROR: %N:%l: on_data_available() -")
               ACE_TEXT(" _narrow failed!\n")));
    ACE_OS::exit(1);
  }

  Messenger::Message message;
  DDS::SampleInfo info;
  while(true){
    const DDS::ReturnCode_t error = reader_i->read_next_sample(message, info);
    std::this_thread::yield();
    if (error == DDS::RETCODE_OK) {
      if (info.valid_data) {
        std::cout << "Message: index    = " << message.index << std::endl
          << "         timestamp = " << message.timestamp << std::endl
          << "         text       = " << message.text.in() << std::endl
          << "         topic = "<< reader_i->get_topicdescription()->get_name() << std::endl;
      }

    }
    else break;
  }
}

void DataReaderListenerImpl::on_subscription_matched(
  DDS::DataReader_ptr reader,
  const DDS::SubscriptionMatchedStatus& /*status*/)
{
  std::cout<<"subscription matched. topic:"<<reader->get_topicdescription()->get_name()<<std::endl;
}
```

subscriber output：
create topic test_topic_1  domain id:10
create topic test_topic_2  domain id:10
subscription matched. topic:test_topic_1
subscription matched. topic:test_topic_2
liveliness changed.topic:test_topic_2 alive_count=1 alive_count_change=1 not_alive_count=0
Message: index    = 1
         timestamp = 1695182583750148
         text       = sdafgdfasdasda
         topic = test_topic_2
Message: index    = 1
         timestamp = 1695182583750127
         text       = sdafgdfasdasda
         topic = test_topic_2

Why is only one topic binding successful, and this topic can also receive data sent by another topic? Is there a problem with my code? Shouldn't topic1 only receive data sent by topic1?

Looking forward to your reply.

**Reply by jrw972** [reply]
To avoid memory leaks, topic and type names should be assigned to a CORBA::String_var.

```
CORBA::String_var topic_name = reader->get_topicdescription()->get_name();
std::cout<<"subscription matched. topic:"<< topic_name.in() <<std::endl;
```

Unless you don't care about dropping samples, you should turn on reliability for your readers.

To answer your question, the correctness of your code depends on `get_writers_key`, which you have not provided.  If this incorrectly returns the same writer, then this would explain your results.  If you want to check this, then

```
std::cout << "Message: index    = " << message.index << std::endl
          << "         timestamp = " << message.timestamp << std::endl
          << "         text       = " << message.text.in() << std::endl
          << "         publication = " << info.publication_handle << std::endl
          << "         topic = "<< reader_i->get_topicdescription()->get_name() << std::endl;
```

Different writers will have different publication handles.

**Comment by zhuohuawang** [comment]
I checked get_writers_key and found nothing wrong with it.
```
//std::unordered_map<std::string, DDS::DataWriter_var> writers_;
std::string Publisher::get_writers_key(int32_t domain_id, std::string topic_name){
    return std::to_string(domain_id) + "/" + topic_name;
}
```

The publication_handle is the same.

**Comment by sonndinh** [comment]
Looks like there is a logic bug in the `Publisher::publish` function. My guess is `message_writer_` is a member data of the `Publisher` class and it's null initially. The first time the function is called, `message_writer_` is assigned to the first writer (to the first topic) and becomes non-null. The second time the function is called, `message_writer_` is not assigned to the second writer (to the second topic) but still is the first writer. So you're actually sending 2 samples of the same topic, and only 1 reader received both samples.

**Reply by zhuohuawang** [reply]
Thank you for your reply. You're right.  

I have another problem: If I change the way the Publisher is created to look like this, then I find that subscribers can only receive one topic. Does this mean that only one participant can be created for the same domainid? 

In fact, I started Publisher this way from the very beginning, until I realized that subscribers could only receive data for one topic. Now I will try to correct the logical error you mentioned, but the result is still the same.

```
std::atomic_bool quit = false;
    for(size_t i =0;i < thrds.size();++i){
        std::thread t([&quit, i](){
            //Create a Publisher for each thread
            Publisher pub;
            std::string topic_name = "test_topic_" + std::to_string(i+1);
            uint32_t domian_id = 10;
            pub.create_publisher(domian_id,topic_name);
            Messenger::Message message;
            message.text = "sdafgdfasdasda";
            for(int index = 0;(index < 100000 && !quit);++index){
                message.index = index;
                message.timestamp = getTimestamp();
                pub.publish(domian_id,topic_name,true,message);
                std::this_thread::sleep_for(std::chrono::milliseconds(5 * 1000));
            }
        });
        thrds[i].swap(t);
    }
```

**Reply by mitza-oci** [reply]
> I have another problem

Each time you have a new topic to discuss, please create a new discussion in GitHub.  The goal of the discussions is to generate content that can be searched and reused by others.

**Reply by zhuohuawang** [reply]
Thank you for reminding me, I'll pay attention next time. But sometimes a different topic can lead to a lack of context.

**Reply by mitza-oci** [reply]
>  sometimes a different topic can lead to a lack of context.

That's what links are for
